test_agent/rust/src/utils.rs

```diff
@@ -0,0 +1,528 @@
+    };
+
+    Ok(UUID {
+        lsb,
+        msb,
+        special_fields: SpecialFields::default(),
+    })
+}
+
+#[derive(Default)]
+pub struct WrapperUPayload(pub UPayload);
+impl<'de> Deserialize<'de> for WrapperUPayload {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let value: Value = Deserialize::deserialize(deserializer)?;
+        let Ok(upayload) = parse_upayload(&value) else {
+            let err_msg = "Error parsing payload: ".to_string();
+            return Err(serde::de::Error::custom(err_msg));
+        };
+
+        Ok(WrapperUPayload(upayload))
+    }
+}
+
+#[allow(clippy::map_unwrap_or)]
+fn parse_upayload(value: &Value) -> Result<UPayload, serde_json::Error> {
+
+    let format = value
+        .get("format")
+        .and_then(|format_value| format_value.as_str())
+        .map(|format_str| {
+            UPayloadFormat::from_str(format_str).unwrap_or_else(|| {
+                error!("Error: Unable to parse string to UPayloadFormat");
+                UPayloadFormat::UPAYLOAD_FORMAT_UNSPECIFIED
+            })
+        })
+        .unwrap_or_else(|| {
+            error!("Error: value of format is not a string");
+            UPayloadFormat::UPAYLOAD_FORMAT_UNSPECIFIED
+        });
+
+    let length = value
+        .get("length")
+        .and_then(Value::as_str)
+        .and_then(|length_value| length_value.parse::<i32>().ok());
+
+    let data = if let Some(data_value) = value.get("value") {
+        if let Ok(data_vec) = serde_json::to_vec(data_value) {
+            Some(Data::Value(data_vec))
+        } else {
+            Some(Data::Reference(0))
+        }
+    } else {
+        return Err(serde::de::Error::custom("Missing value field"));
+    };
+
+    Ok(UPayload {
+        format: format.into(),
+        length,
+        data,
+        special_fields: SpecialFields::default(),
+    })
+}
+
+#[derive(Debug, Default)]
+
+pub struct WrapperUMessage(pub UMessage);
+
+impl<'de> Deserialize<'de> for WrapperUMessage {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let value: Value = Deserialize::deserialize(deserializer)?;
+
+        let wattributes = match value.get("attributes") {
+            Some(attributes) => {
+                serde_json::from_value::<WrapperUAttribute>(attributes.clone()).unwrap_or_default()
+            }
+            None => WrapperUAttribute(UAttributes::default()),
+        };
+
+        let wpayload = match value.get("payload") {
+            Some(payload) => {
+                serde_json::from_value::<WrapperUPayload>(payload.clone()).unwrap_or_default()
+            }
+            None => WrapperUPayload(UPayload::default()),
+        };
+
+        Ok(WrapperUMessage(UMessage {
+            attributes: Some(wattributes.0).into(),
+            payload: Some(wpayload.0).into(),
+            special_fields: SpecialFields::default(),
```

Comment:
Same comments about special fields as elsewhere. Avoid em by using Default trait.

Reply:
same as https://rust-lang.github.io/rust-clippy/master/index.html#/default_trait_access

Reply:
Check out my response on the other comment. Still suggest doing this.

Reply:
The `..` thing at the end of initializing a struct is called [struct update syntax](https://google.github.io/comprehensive-rust/std-traits/default.html) and the linked page does a pretty good job of showing the power of using this together with the Default trait implementation _of the struct being initialized_.